projects/rocthrust/thrust/unique_ptr.h

```diff
@@ -0,0 +1,551 @@
+#pragma once
+
+#include <thrust/detail/config.h>
+
+#include <thrust/detail/type_traits.h>
+#include <thrust/device_free.h>
+#include <thrust/device_new.h>
+#include <thrust/device_ptr.h>
+#include <thrust/device_reference.h>
+#include <thrust/for_each.h>
+
+#include <compare>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+THRUST_NAMESPACE_BEGIN
+
+template <class T, class = void>
+struct default_delete;
+
+template <class T>
+struct default_delete<T, typename std::enable_if<!std::is_array<T>::value>::type>
+{
+  using pointer = thrust::device_ptr<T>;
+
+  THRUST_HOST constexpr default_delete() noexcept = default;
+
+  template <class U>
+  THRUST_HOST default_delete(
+    const default_delete<U>&,
+    typename std::enable_if<std::is_convertible<thrust::device_ptr<U>, pointer>::value>::type* = nullptr) noexcept
```

Comment:
Nit, feel free to ignore. But we can now use C++17, so it could a bit more readable if we use the `_t` and `_v` variants of the type traits.

```suggestion
    typename std::enable_if_t<std::is_convertible_v<thrust::device_ptr<U>, pointer>>* = nullptr) noexcept
```

Applies to other places too, but is basically a form of find and replace.

Reply:
Agree, I'll make that change. 

When we started developing this we used `thrust::detail` type traits, we didn't know that a switch to C++17 type traits was going to happen and that's why it's like this.